pkg/cache/cache.go

```diff
@@ -249,6 +249,12 @@ type ByObject struct {
 	// Be very careful with this, when enabled you must DeepCopy any object before mutating it,
 	// otherwise you will mutate the object in the cache.
 	UnsafeDisableDeepCopy *bool
+
+	// NewCache allows to use a custom cache creation function for the object.
+	//
+	// NOTE: LOW LEVEL PRIMITIVE!
+	// Only use a custom NewCache if you know what you are doing.
+	NewCache NewCacheFunc
```

Comment:
Few point:
- Setting the cache using `NewCache` means all comments about other values in ByObject will become irrelevant. Please adapt those doc strings. 
- If `NewCache` is there we should not allow setting other fields of `ByObject`.
- [Defaulting](https://github.com/kubernetes-sigs/controller-runtime/blob/c93e2abcb28eb71fccad7dc565f0547cc07e5566/pkg/cache/cache.go#L405-L442) should not be done in case of `NewCache` is set for `ByObject`.

Reply:
I wonder about all of the above points. Shouldn't we rather create a copy of `Options` and remove all `ByObject` items that are not relevant for this object type? We don't now what information `NewCache` uses.

Reply:
Please see [cc3b976](https://github.com/kubernetes-sigs/controller-runtime/pull/2518/commits/cc3b9767fa293af93f6349779b7cfac20a878477) and tell me what you think (cc @troy0820) 🙂